sysrib/server_test.go

```diff
@@ -0,0 +1,653 @@
+package sysrib
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/openconfig/gribigo/afthelper"
+	pb "github.com/openconfig/lemming/proto/sysrib"
+)
+
+type AddIntfAction struct {
+	name    string
+	ifindex int32
+	enabled bool
+	prefix  string
+	niName  string
+}
+
+type SetRouteRequestAction struct {
+	Desc     string
+	RouteReq *pb.SetRouteRequest
+}
+
+type FakeDataplane struct {
+	incomingRoutes []*ResolvedRoute
+}
+
+func (dp *FakeDataplane) ProgramRoute(r *ResolvedRoute) error {
+	dp.incomingRoutes = append(dp.incomingRoutes, r)
+	// Assume all routes are programmed successfully.
+	return nil
+}
+
+func (dp *FakeDataplane) GetRoutesAndClearQueue() []*ResolvedRoute {
+	rs := dp.incomingRoutes
+	dp.incomingRoutes = []*ResolvedRoute{}
+	return rs
+}
+
+func NewFakeDataplane() *FakeDataplane {
+	return &FakeDataplane{}
+}
+
+// routeSliceToMap converts a slice of ResolvedRoute to a map keyed by their
+// RouteKeys. It returns an error if any of the routes were nil or if there is
+// a duplicate.
+func routeSliceToMap(rs []*ResolvedRoute) (map[RouteKey]*ResolvedRoute, error) {
+	ret := map[RouteKey]*ResolvedRoute{}
+	for _, rr := range rs {
+		if rr == nil {
+			return nil, fmt.Errorf("Got nil route in ResolvedRoute slice")
+		}
+		if existing, ok := ret[rr.RouteKey]; ok {
+			return nil, fmt.Errorf("Got duplicate route key:\nFirst: %+v\nDuplicate: %+v", existing, rr)
+		}
+		ret[rr.RouteKey] = rr
+	}
+	return ret, nil
+}
+
+func checkResolvedRoutesEqual(got, want []*ResolvedRoute) error {
+	gotRoutes, err := routeSliceToMap(got)
```

Comment:
(nit) you can t.Helper() here so that the function doesn't get reported as the source of the failure.

Reply:
Thanks, I'm actually not passing testing.T here because I wanted the downstream to report the errors with some context instead.